bindings_generator/src/special_methods.rs

```diff
@@ -135,243 +150,245 @@ pub fn generate_free_impl(
         let snake_name = class_name_to_snake_case(&base_class_name);
-        if is_pointer_safe {
-            writeln!(
-                output,
-                r#"    /// Up-cast.
-    #[inline]
-    pub fn to_{snake_name}(&self) -> {name} {{
-        {addref_if_reference}
-        unsafe {{ {name}::from_sys(self.this) }}
-    }}"#,
-                name = parent.name,
-                snake_name = snake_name,
-                addref_if_reference = if parent.is_refcounted() {
-                    "unsafe {{ object::add_ref(self.this); }}"
-                } else {
-                    "// Not reference-counted."
-                },
-            )?;
+        let parent_class = format_ident!("{}", parent.name);
+        let to_snake_name = format_ident!("to_{}", snake_name);
+        let addref_if_reference = if parent.is_refcounted() {
+            quote! {
+                unsafe { object::add_ref(self.this); }
+            }
+        } else {
+            quote! {
+                // Not reference-counted.
+            }
+        };
+        let maybe_unsafe = if is_pointer_safe {
+            Default::default()
         } else {
-            writeln!(
-                output,
-                r#"
-    /// Up-cast.
-    #[inline]
-    pub unsafe fn to_{snake_name}(&self) -> {name} {{
-        {addref_if_reference}
-        unsafe {{ {name}::from_sys(self.this) }}
-    }}"#,
-                name = parent.name,
-                snake_name = snake_name,
-                addref_if_reference = if parent.is_refcounted() {
-                    "object::add_ref(self.this);"
-                } else {
-                    "// Not reference-counted."
-                },
-            )?;
+            quote! { unsafe }
+        };
+
+        let upcast = generate_upcast(api, &parent.base_class, is_pointer_safe);
+        quote! {
+            /// Up-cast.
+            #[inline]
+            pub #maybe_unsafe fn #to_snake_name(&self) -> #parent_class {
+                #addref_if_reference
+                unsafe { #parent_class::from_sys(self.this) }
+            }
+
+            #upcast
         }
-
-        generate_upcast(output, api, &parent.base_class, is_pointer_safe)?;
+    } else {
+        Default::default()
     }
-
-    Ok(())
 }
 
 pub fn generate_deref_impl(output: &mut impl Write, class: &GodotClass) -> GeneratorResult {
-    writeln!(
-        output,
-        r#"
-impl std::ops::Deref for {name} {{
-    type Target = {base};
-
-    #[inline]
-    fn deref(&self) -> &{base} {{
-        unsafe {{
-            std::mem::transmute(self)
-        }}
-    }}
-}}
-
-impl std::ops::DerefMut for {name} {{
-    #[inline]
-    fn deref_mut(&mut self) -> &mut {base} {{
-        unsafe {{
-            std::mem::transmute(self)
-        }}
-    }}
-}}"#,
-        name = class.name,
-        base = class.base_class,
-    )?;
+    if class.base_class.is_empty() {
+        return Ok(());
+    }
```

Comment:
This should just be an assert. It's better to add a branch where this is called.